File: programmers/kakao/level4/호텔_방_배정.py
def solution(k, room_number):
    answer = []
    rooms = {}
    for num in room_number:
        n = num
        visited = [n]
        while n in rooms:
            n = rooms[n]
            visited.append(n)
        answer.append(n)
        for v in visited:
            rooms[v] = n+1
    return answer


k, room_number = 10, [1,3,4,1,3,1]
print(solution(k, room_number)) # [1,3,4,2,5,6]

File: Java/programmers/kakao/level4/호텔방배정.java
package CodeTest.Java.programmers.kakao.level4;

import java.util.*;

public class 호텔방배정 {
    private static Map<Long, Long> rooms;

    public long[] solution(long k, long[] room_number) {
        int n = room_number.length;

        long[] answer = new long[n];
        rooms = new HashMap<>();

        for (int i=0; i<n; i++) {
            answer[i] = find(room_number[i]);
        }

        return answer;
    }

    private static long find(long num) {
        if (!rooms.containsKey(num)) {
            rooms.put(num, num+1);
            return num;
        }

        long empty = find(rooms.get(num));
        rooms.put(num, empty+1);
        
        return empty;
    }
}
